### PR TITLE
Rewrite camera support for the Espressif port

### PR DESCRIPTION
**Pull request by jepler**
This removes the old camera module (imagecapture) from the Espressif port and replaces it with a wrapper around esp32-camera, a very capable library that handles a whole range of cameras on esp32, esp32-s2 and esp32-s3.

The old Python modules such as adafruit_circuitpython_ov5640 are not used anymore, as the C code does mode setting. The API is also unabashedly different than the old Python modules, and is instead intended to be more like the esp32-camera C API.

Creating the PR in draft form because there are TODOs and known problems:

- [x] Enable code to work across soft reset (exception running code a second time) and `deinit()` (crashes). Worryingly, the public API `esp_camera_deinit` is noted as not working! See e.g., https://github.com/espressif/esp32-camera/issues/426 https://github.com/espressif/esp32-camera/issues/54 and possibly others
- [x] Implement properties for camera settings (contrast, brightness, etc)
- [x] Test pattern support seems to be missing in the esp32-camera API -- no, it's called `set_colorbar`
- [x] Enable the LCD on the ESP32-S3-EYE
- [x] Add the ESP32-EYE
- [x] fails (times out) taking 2nd picture, at least on ESP32-EYE (local non-pushed changes add this board)

If any of the following turn up additional changes, further PR(s) can be created:
- Further testing & robustness fixes as needed
- Prepare a pull request into esp32-camera with our enhancements, and find out if they're desirable
- Make versions of existing demos that work with this module
- Add any missing functionality identified during the above process

This PR also adds support for the ESP32-S3-EYE devekit (pending allocation of PIDs for it: https://github.com/espressif/usb-pids/pull/54)

This is tested on the Kaluga (ESP32-S2), ESP32-EYE, and and ESP32-S3-EYE boards. The Kaluga has an OV5640 and the EYEs have an OV2640.

First, reserve some PSRAM by putting this line in `CIRCUITPY/_env`:
```
CIRCUITPY_RESERVED_PSRAM=524288
```
and hard-reset the board for it to take effect. (EYEs reserve 1MB by default, so this is not needed there)

Now, the following script will take a very low-resolution jpeg file and print it in the REPL in escape coded form:

```python
import board
import esp32_camera

c = esp32_camera.Camera(
    data_pins=board.CAMERA_DATA,
    external_clock_pin=board.CAMERA_XCLK,
    pixel_clock_pin=board.CAMERA_PCLK,
    vsync_pin=board.CAMERA_VSYNC,
    href_pin=board.CAMERA_HREF,
    pixel_format=esp32_camera.PixelFormat.JPEG,
    i2c=board.I2C(),
    external_clock_frequency=20_000_000)

m = c.take()
if m is not None:
    print(bytes(m))
```

Then on desktop open a python repl and run something like
```python
>>> with open("my.jpg", "wb") as f: f.write(<BIG PASTE FROM REPL>)
```
and open my.jpg in a viewer.

### CODE REVIEW

**Comment by dhalbert** [comment]
I wonder if the idea of reserving RAM for some purpose or other could be generalized across ports. The reserved chunk could be available from, say, `supervisor` as a bytearray you could pass somewhere, such as to the camera API. Or does the library need to know exactly where the reserved RAM is?

**Comment by jepler** [comment]
Yeah, the problem that I'm 'attacking' with this patch is that esp-idf components such as camera and lcd want to allocate things only through esp-idf allocation APIs. We, of course, chose to use the whole psram region for the CircuitPython heap (and "supervisor allocations", so it's unavailable to the system allocator.

(wifi is of course the same, except that its allocations are modestly sized compared to the framebuffers that a camera works with)

What I found was that there's an API in esp-idf that can be used to grant an arbitrary region to the esp-idf allocation APIs, forever. However, this is using a unique esp-idf API to grant the region; and the region can never be "un-granted".

I'm not sure how it would be generalized to other ports.

**Comment by jepler** [comment]
@dhalbert any advice for the ports that are out of space with this addition?  I can disable the module on funhouse, magtag and mixgo since they don't have enough exposed pins anyway, but technically I think a feather could have a camera module attached and not quite run out of GPIOs.

**Comment by jepler** [comment]
the feathers that failed have already had -Os applied so that trick won't help.

**Comment by dhalbert** [comment]
The space that @tannewt recovered with the -Os libc I think will fix the `arduino_nano_33_iot` build? `ru` is overflowing slightly.

File: ports/espressif/bindings/esp32_camera/Camera.c [hunk 1]
```diff
@@ -0,0 +1,1017 @@
...
+    check_for_deinit(self);
+    return cp_enum_find(&esp32_camera_frame_size_type, common_hal_esp32_camera_camera_get_frame_size(self));
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(esp32_camera_camera_get_frame_size_obj, esp32_camera_camera_get_frame_size);
+
+MP_PROPERTY_GETTER(esp32_camera_camera_frame_size_obj,
+    (mp_obj_t)&esp32_camera_camera_get_frame_size_obj);
+
+//|     contrast: int
+//|     """Access the contrast property of the camera sensor"""
```
Comment (tannewt) [inline ports/espressif/bindings/esp32_camera/Camera.c:307]: I know it's annoying but these properties should really explain what the property is. That way folks don't need to look further for that info. If a property is unlikely to be used, then just remove it. It is ok to not expose everything.

**Review by tannewt — CHANGES_REQUESTED** [review]
Overall, this looks pretty good. I'd like to see a bit more info and discretion about the properties of the camera that are exposed.

File: ports/espressif/boards/espressif_esp32_eye/mpconfigboard.mk [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+CIRCUITPY_CREATION_ID = 0x00320001
+
+IDF_TARGET = esp32
+
+INTERNAL_FLASH_FILESYSTEM = 1
+LONGINT_IMPL = MPZ
+
+CIRCUITPY_STATUS_BAR = 0
+CIRCUITPY_WEB_WORKFLOW = 0
+
```
[the remark this answers is not in the document]
Comment (jepler) [reply ports/espressif/boards/espressif_esp32_eye/mpconfigboard.mk:11]: mistake, will revert.

File: ports/espressif/bindings/esp32_camera/Camera.c (hunk 1 above)
Comment (jepler) [reply ports/espressif/bindings/esp32_camera/Camera.c:307]: yes, agreed.

**Review by tannewt — APPROVED** [review]
Much better property descriptions! Thank you!

**Review by tannewt — APPROVED** [review]

**Comment by ladyada** [comment]
@jepler hiya just to check - because of https://github.com/espressif/esp32-camera/issues/433#issuecomment-1204547256 we can't switch between viewfinder and jpeg mode 'dynamically'?

**Comment by jepler** [comment]
@ladyada we work around it for ourselves, I tested my workaround for the limitation but none of the examples I committed use this facility.

Basically you call reconfigure on the camera object https://github.com/jepler/circuitpython/blob/98202c85688384f702802ad263c30afa8c74da9e/ports/espressif/bindings/esp32_camera/Camera.c#L227

**Comment by ladyada** [comment]
kool will experiment!

**Comment by jepler** [comment]
@ladyada I committed a demo that shows live view on lcd, then uploads to adafruit io when the BOOT button is clicked. https://github.com/jepler/Adafruit_Learning_System_Guides/blob/456ab88f211dfbf4533c8aed9586f3db4176b03d/circuitpython-esp32-camera/esp32-s3-eye-adafruitio-and-lcd/code.py it'll go in the eventual code for learn

**Pull request merged**